crates/pico_macros/src/source.rs

```diff
@@ -0,0 +1,58 @@
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::{parse_macro_input, DeriveInput};
+
+pub(crate) fn source(_args: TokenStream, item: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(item as DeriveInput);
+
+    let struct_name = input.clone().ident;
+    let key = struct_name.to_string();
+
+    let output = quote! {
+        #[derive(Debug, Clone, PartialEq, Eq)]
+        #input
+
+        impl #struct_name {
+            pub fn set(
+                self,
+                db: &mut pico::database::Database,
+                static_key: &'static str,
```

Comment:
can you explain what `static_key` is doing? Is this some stable key for a given source node?

I would hope that this can be transparent somehow. Perhaps require that `Input` derive a trait, and have a `static_key` fn on that trait?

Reply:
for a source it could be a filename or iso literal name. Salsa allows to put #[id] macro on some unique field in the struct. We can go this way as well and require static_key to be a part of a struct

Reply:
IMO a method is the best in the long run (or if there are non-Isograph users of pico), so that it can be derived from other fields, but yeah, if we want to make it a required field, that's fine too! We can always change it later